src/internal/starlark/startest/startest.go

```diff
@@ -0,0 +1,210 @@
+			args = append(args, originalArgs[1:]...)
+		} else {
+			msg = "%v: invalid format string: "
+			args = append(args, originalArgs...)
+		}
+	}
+	f(msg, args...)
+}
+
+func caller(th *starlark.Thread) string {
+	stack := th.CallStack()
+	caller := strings.ReplaceAll(stack.String(), "\n", "/") // emergency backup
+	if len(stack) > 1 {
+		i := len(stack) - 2
+		caller = fmt.Sprintf("%v:%v:%v", stack[i].Pos.Filename(), stack[i].Pos.Line, stack[i].Pos.Col)
+	}
+	return caller
+}
+
+func makeStarlarkTestFunction(name string, f func(t T, caller string, args []any)) *starlark.Builtin {
+	return starlark.NewBuiltin(name, func(th *starlark.Thread, fn *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+		t := getTestingT(th)
+		caller := caller(th)
+		if len(kwargs) != 0 {
+			t.Errorf("%v: unexpected kwargs in %v", caller, fn.Name())
+			return starlark.None, nil
+		}
+		var goArgs []any
+		for _, arg := range args {
+			v := ourstar.FromStarlark(arg)
+			goArgs = append(goArgs, v)
+		}
+		f(t, caller, goArgs)
+		return starlark.None, nil
+	})
+}
+
+func getTestingT(th *starlark.Thread) T {
+	return th.Local(testingTKey).(T)
+}
+
+func setTestingT(th *starlark.Thread, t T) {
+	th.SetLocal(testingTKey, t)
+}
+
+func setupPrint(th *starlark.Thread, t T) {
+	// Redirect print() to t.Log.
+	th.Print = func(th *starlark.Thread, msg string) {
+		t.Logf("%v: %v", caller(th), msg)
+	}
+}
+
+// RunTest runs an on-disk Starlark test.  Tests are scripts that produce functions that start with
+// the name "test_", and take a *testing.T object as their first argument.  The tests work like Go
+// tests; t.Error fails the test and continues, t.Fatal aborts the test.  Each Starlark test
+// function is run as a subtest of a "starlark" test created by this command.  Starlark tests cannot
+// call t.Parallel, since we use the same Starlark thread for each test, but there is no reason why
+// this stipulation cannot be removed.
+func RunTest(t *testing.T, script string, opts ourstar.Options) {
+	ctx := pctx.TestContext(t)
+	runTest(ctx, &testingTWrapper{T: t}, script, opts)
+}
+
+func runTest(ctx context.Context, t T, script string, opts ourstar.Options) {
+	t.Helper()
+	t.Run("starlark", func(t T) {
+		if _, err := ourstar.Run(ctx, script, opts, func(fileOpts *syntax.FileOptions, th *starlark.Thread, in, module string, globals starlark.StringDict) (starlark.StringDict, error) {
+			setupPrint(th, t)
+			result, err := starlark.ExecFileOptions(fileOpts, th, module, nil, globals)
+			if err != nil {
+				return nil, errors.Wrap(err, "compile test")
+			}
+
+			// We can't easily investigate the top-level identifiers in source code
+			// order; too much information from the parser fails to make it here.  So,
+			// we look for valid tests in alphabetical order, and then run those tests
+			// in source code order.
+			identifiers := maps.Keys(result)
+			sort.Strings(identifiers)
```

Comment:
why this sort?

Reply:
This is for consistently reporting the error messages that appear in invalid test files; they would be in arbitrary order since ExecFileOptions returns the globals of the file in a map without source location information.  Once we have filtered this down to valid test functions, though, we can inspect the function object to find the source code position and actually run valid tests in the order they appear in the source code.